File: gemini/typescript/examples/cached-content.ts
/**
 * Cached Content Example
 *
 * This example demonstrates:
 * - Creating cached content with TTL
 * - Using cached content in generation requests
 * - Updating cached content expiration
 * - Listing cached content
 * - Deleting cached content
 *
 * Cached content allows you to cache large context data (like documents)
 * and reuse it across multiple generation requests, reducing costs and latency.
 *
 * Prerequisites:
 * - Set the GEMINI_API_KEY environment variable
 *
 * Usage:
 * ```bash
 * export GEMINI_API_KEY="your-api-key"
 * npm run build
 * node dist/examples/cached-content.js
 * ```
 */

import { createClientFromEnv } from '../src/index.js';

/**
 * Example 1: Create cached content with TTL
 */
async function exampleCreateCachedContent(): Promise<string> {
  console.log('\n=== Example 1: Create Cached Content ===\n');

  const client = createClientFromEnv();

  // Create a large context to cache
  const systemInstruction = `You are an expert in quantum physics. You have deep knowledge of:
- Quantum mechanics principles
- Wave-particle duality
- Quantum entanglement
- Superposition
- Quantum computing
- Schrödinger's equation
- Heisenberg uncertainty principle

When answering questions, be precise, technical, and educational.`;

  const contextDocument = `# Introduction to Quantum Computing

Quantum computing is a revolutionary approach to computation that leverages quantum mechanical phenomena
such as superposition and entanglement to process information. Unlike classical computers that use bits
(0 or 1), quantum computers use quantum bits or qubits that can exist in multiple states simultaneously.

## Key Principles

### Superposition
A qubit can be in a state of 0, 1, or any quantum superposition of these states. This allows quantum
computers to process vast amounts of information simultaneously.

### Entanglement
Quantum entanglement is a phenomenon where qubits become correlated in such a way that the state of
one qubit depends on the state of another, regardless of the distance between them.

### Quantum Gates
Quantum gates manipulate qubits to perform calculations. Unlike classical logic gates, quantum gates
are reversible and operate on the probability amplitudes of quantum states.

## Applications
- Cryptography and security
- Drug discovery and molecular modeling
- Optimization problems
- Machine learning and AI
- Financial modeling`;

  console.log('Creating cached content with system instruction and context...');
  console.log(`TTL: 3600 seconds (1 hour)\n`);

  const cachedContent = await client.cachedContent.create({
    model: 'models/gemini-2.0-flash-exp',
    contents: [
      {
        role: 'user',
        parts: [{ text: contextDocument }],
      },
    ],
    systemInstruction: {
      parts: [{ text: systemInstruction }],
    },
    ttl: '3600s', // Cache for 1 hour
  });

  console.log('Cached content created successfully!');
  console.log(`Name: ${cachedContent.name}`);
  console.log(`Model: ${cachedContent.model}`);
  console.log(`Created: ${cachedContent.createTime}`);
  console.log(`Updated: ${cachedContent.updateTime}`);
  console.log(`Expires: ${cachedContent.expireTime}`);

  if (cachedContent.usageMetadata) {
    console.log(`\nUsage Metadata:`);
    console.log(`  Total Token Count: ${cachedContent.usageMetadata.totalTokenCount}`);
  }

  return cachedContent.name;
}

/**
 * Example 2: Use cached content in generation
 */
async function exampleUseCache(cacheName: string): Promise<void> {
  console.log('\n=== Example 2: Generate Content Using Cache ===\n');

  const client = createClientFromEnv();

  console.log(`Using cached content: ${cacheName}`);
  console.log('Generating response to query about quantum computing...\n');

  // Use the cached content in a generation request
  // The cache name is passed in the GenerateContentRequest
  const response = await client.content.generate('gemini-2.0-flash-exp', {
    contents: [
      {
        parts: [
          {
            text: 'Based on the context provided, explain quantum entanglement in simple terms.',
          },
        ],
      },
    ],
    // Reference the cached content by name
    cachedContent: cacheName,
  });

  if (response.candidates && response.candidates.length > 0) {
    const text = response.candidates[0].content?.parts
      .map((part) => ('text' in part ? part.text : ''))
      .join('');

    console.log('Generated Response:');
    console.log(text);

    // Show usage metadata
    if (response.usageMetadata) {
      console.log('\n--- Token Usage ---');
      console.log(`Prompt Tokens: ${response.usageMetadata.promptTokenCount}`);
      console.log(`Candidates Tokens: ${response.usageMetadata.candidatesTokenCount}`);
      console.log(`Total Tokens: ${response.usageMetadata.totalTokenCount}`);

      // Cached content usage
      if (response.usageMetadata.cachedContentTokenCount) {
        console.log(`Cached Content Tokens: ${response.usageMetadata.cachedContentTokenCount}`);
        console.log('(These tokens were served from cache, reducing cost!)');
      }
    }
  }
}

/**
 * Example 3: List cached contents
 */
async function exampleListCachedContent(): Promise<void> {
  console.log('\n=== Example 3: List Cached Contents ===\n');

  const client = createClientFromEnv();

  console.log('Listing all cached contents...\n');

  const response = await client.cachedContent.list({
    pageSize: 10,
  });

  if (response.cachedContents && response.cachedContents.length > 0) {
    console.log(`Found ${response.cachedContents.length} cached content(s):\n`);

    for (const cache of response.cachedContents) {
      console.log(`- ${cache.name}`);
      console.log(`  Model: ${cache.model}`);
      console.log(`  Created: ${cache.createTime}`);
      console.log(`  Expires: ${cache.expireTime}`);

      if (cache.usageMetadata) {
        console.log(`  Tokens: ${cache.usageMetadata.totalTokenCount}`);
      }

      console.log();
    }

    if (response.nextPageToken) {
      console.log('More cached contents available. Use nextPageToken to fetch next page.');
    }
  } else {
    console.log('No cached contents found.');
  }
}

/**
 * Example 4: Get cached content metadata
 */
async function exampleGetCachedContent(cacheName: string): Promise<void> {
  console.log('\n=== Example 4: Get Cached Content Metadata ===\n');

  const client = createClientFromEnv();

  console.log(`Fetching metadata for: ${cacheName}\n`);

  const cachedContent = await client.cachedContent.get(cacheName);

  console.log('Cached content details:');
  console.log(`Name: ${cachedContent.name}`);
  console.log(`Model: ${cachedContent.model}`);
  console.log(`Display Name: ${cachedContent.displayName || 'N/A'}`);
  console.log(`Created: ${cachedContent.createTime}`);
  console.log(`Updated: ${cachedContent.updateTime}`);
  console.log(`Expires: ${cachedContent.expireTime}`);

  if (cachedContent.usageMetadata) {
    console.log(`\nUsage Metadata:`);
    console.log(`  Total Token Count: ${cachedContent.usageMetadata.totalTokenCount}`);
  }

  console.log(`\nSystem Instruction:`);
  if (cachedContent.systemInstruction) {
    const sysText = cachedContent.systemInstruction.parts
      .map((part) => ('text' in part ? part.text : ''))
      .join('');
    console.log(sysText.substring(0, 150) + '...');
  }

  console.log(`\nContents: ${cachedContent.contents.length} item(s)`);
}

/**
 * Example 5: Update cached content expiration
 */
async function exampleUpdateCachedContent(cacheName: string): Promise<void> {
  console.log('\n=== Example 5: Update Cached Content Expiration ===\n');

  const client = createClientFromEnv();

  console.log(`Updating expiration for: ${cacheName}`);
  console.log('Extending TTL by 1 hour...\n');

  const updatedCache = await client.cachedContent.update(cacheName, {
    ttl: '7200s', // 2 hours
  });

  console.log('Cached content updated successfully!');
  console.log(`Name: ${updatedCache.name}`);
  console.log(`Updated: ${updatedCache.updateTime}`);
  console.log(`New Expiration: ${updatedCache.expireTime}`);
}

/**
 * Example 6: Create cached content with absolute expiration time
 */
async function exampleCreateWithExpireTime(): Promise<string | null> {
  console.log('\n=== Example 6: Create with Absolute Expiration Time ===\n');

  const client = createClientFromEnv();

  // Set expiration to 2 hours from now
  const expireTime = new Date(Date.now() + 2 * 60 * 60 * 1000);

  console.log('Creating cached content with absolute expiration time...');
  console.log(`Expires at: ${expireTime.toISOString()}\n`);

  try {
    const cachedContent = await client.cachedContent.create({
      model: 'models/gemini-2.0-flash-exp',
      contents: [
        {
          parts: [{ text: 'This is a short cached context for demonstration.' }],
        },
      ],
      expireTime: expireTime.toISOString(),
    });

    console.log('Cached content created successfully!');
    console.log(`Name: ${cachedContent.name}`);
    console.log(`Expires: ${cachedContent.expireTime}`);

    return cachedContent.name;
  } catch (error) {
    console.log('Note: This may fail if the model or API does not support absolute expireTime.');
    console.log('Error:', error);
    return null;
  }
}

/**
 * Example 7: Delete cached content
 */
async function exampleDeleteCachedContent(cacheName: string): Promise<void> {
  console.log('\n=== Example 7: Delete Cached Content ===\n');

  const client = createClientFromEnv();

  console.log(`Deleting cached content: ${cacheName}`);

  await client.cachedContent.delete(cacheName);

  console.log('Cached content deleted successfully!');

  // Verify deletion
  try {
    await client.cachedContent.get(cacheName);
    console.log('WARNING: Cached content still exists after deletion!');
  } catch (error) {
    console.log('Verified: Cached content no longer exists.');
  }
}

/**
 * Main function running all examples
 */
async function main(): Promise<void> {
  let cacheName1: string | null = null;
  let cacheName2: string | null = null;

  try {
    console.log('=== Cached Content Examples ===');

    // Create cached content
    cacheName1 = await exampleCreateCachedContent();

    // Use cached content
    await exampleUseCache(cacheName1);

    // List and get
    await exampleListCachedContent();
    await exampleGetCachedContent(cacheName1);

    // Update expiration
    await exampleUpdateCachedContent(cacheName1);

    // Create with absolute expiration time
    cacheName2 = await exampleCreateWithExpireTime();

    console.log('\n=== All examples completed successfully! ===');
  } catch (error) {
    console.error('\nError during cached content operations:', error);
    if (error instanceof Error) {
      console.error('Error message:', error.message);
      console.error('Stack trace:', error.stack);
    }
    process.exit(1);
  } finally {
    // Clean up: Delete cached contents
    console.log('\n=== Cleanup ===\n');

    const client = createClientFromEnv();

    if (cacheName1) {
      try {
        await exampleDeleteCachedContent(cacheName1);
      } catch (error) {
        console.error(`Failed to delete ${cacheName1}:`, error);
      }
    }

    if (cacheName2) {
      try {
        await exampleDeleteCachedContent(cacheName2);
      } catch (error) {
        console.error(`Failed to delete ${cacheName2}:`, error);
      }
    }
  }
}

// Run the examples
main();

File: gemini/rust/examples/cached_content.rs
//! Cached content example for Gemini API.
//!
//! This example demonstrates:
//! - Creating cached content with TTL (time-to-live)
//! - Using cached content in generation requests
//! - Listing cached content
//! - Updating cached content expiration
//! - Deleting cached content
//! - Understanding token savings from caching
//!
//! # Usage
//!
//! Set your API key as an environment variable:
//! ```bash
//! export GEMINI_API_KEY="your-api-key-here"
//! ```
//!
//! Then run:
//! ```bash
//! cargo run --example cached_content
//! ```

use integrations_gemini::{
    GeminiClientImpl,
    types::{
        CreateCachedContentRequest, UpdateCachedContentRequest, ListCachedContentsParams,
        Content, Part, Role, GenerateContentRequest,
    },
};
use std::time::Duration;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize tracing
    tracing_subscriber::fmt()
        .with_max_level(tracing::Level::INFO)
        .with_target(false)
        .init();

    println!("=== Gemini Cached Content Example ===\n");

    // Create client
    println!("1. Creating Gemini client from environment...");
    let client = GeminiClientImpl::from_env()?;
    println!("   ✓ Client created successfully\n");

    // Example 1: Create cached content
    println!("=== Example 1: Create Cached Content ===\n");
    let cached_name = example_create_cached_content(&client).await?;

    println!("\n");

    // Example 2: List cached content
    println!("=== Example 2: List Cached Content ===\n");
    example_list_cached_content(&client).await?;

    println!("\n");

    // Example 3: Use cached content in generation
    println!("=== Example 3: Use Cached Content in Generation ===\n");
    example_use_cached_content(&client, &cached_name).await?;

    println!("\n");

    // Example 4: Update cached content expiration
    println!("=== Example 4: Update Cached Content Expiration ===\n");
    example_update_cached_content(&client, &cached_name).await?;

    println!("\n");

    // Example 5: Delete cached content
    println!("=== Example 5: Delete Cached Content ===\n");
    example_delete_cached_content(&client, &cached_name).await?;

    println!("\n=== Examples Complete ===");

    Ok(())
}

/// Example: Create cached content with TTL.
async fn example_create_cached_content(client: &GeminiClientImpl) -> Result<String, Box<dyn std::error::Error>> {
    println!("2. Creating cached content...");

    // Create a large system instruction that will be cached
    let system_instruction = Content {
        role: Some(Role::System),
        parts: vec![
            Part::Text {
                text: r#"You are a helpful AI assistant specialized in explaining complex topics.

Your responses should:
1. Be clear and concise
2. Use analogies where appropriate
3. Break down complex concepts into digestible parts
4. Provide examples when helpful
5. Adapt your explanation level to the question

When explaining technical concepts:
- Start with the big picture
- Define key terms
- Explain how components interact
- Use real-world analogies
- Avoid jargon unless necessary

Always maintain a friendly and encouraging tone.
"#.to_string(),
            }
        ],
    };

    let request = CreateCachedContentRequest {
        model: Some("models/gemini-1.5-pro".to_string()),
        contents: vec![system_instruction.clone()],
        system_instruction: None,
        tools: None,
        tool_config: None,
        ttl: Some("3600s".to_string()), // 1 hour TTL
        expire_time: None,
        display_name: Some("helpful-assistant-context".to_string()),
    };

    println!("   Model: gemini-1.5-pro");
    println!("   TTL: 3600s (1 hour)");
    println!("   Display name: helpful-assistant-context");
    println!("   System instruction length: {} characters",
             system_instruction.parts.iter()
                 .filter_map(|p| if let Part::Text { text } = p { Some(text.len()) } else { None })
                 .sum::<usize>());

    // Note: CachedContentService not yet implemented
    // Uncomment when available:
    /*
    let cached_content = client.cached_content()
        .create(request)
        .await?;

    println!("\n   ✓ Cached content created successfully");
    println!("   Name: {}", cached_content.name);
    println!("   Create time: {}", cached_content.create_time.unwrap_or_default());
    println!("   Expire time: {}", cached_content.expire_time.unwrap_or_default());

    if let Some(usage) = &cached_content.usage_metadata {
        println!("\n   Token usage:");
        println!("     Total tokens: {}", usage.total_token_count);
    }

    return Ok(cached_content.name);
    */

    println!("\n   Note: CachedContentService implementation is pending.");
    Ok("cachedContents/sample-cache-id".to_string())
}

/// Example: List cached content.
async fn example_list_cached_content(client: &GeminiClientImpl) -> Result<(), Box<dyn std::error::Error>> {
    println!("2. Listing cached content...");

    let params = ListCachedContentsParams {
        page_size: Some(10),
        page_token: None,
    };

    // Note: CachedContentService not yet implemented
    // Uncomment when available:
    /*
    let response = client.cached_content()
        .list(Some(params))
        .await?;

    println!("\n   ✓ Cached content retrieved successfully");

    if let Some(cached_contents) = &response.cached_contents {
        println!("   Found {} cached content(s):", cached_contents.len());

        for (i, cached) in cached_contents.iter().enumerate() {
            println!("\n   Cached Content {}:", i + 1);
            println!("     Name: {}", cached.name);
            println!("     Display name: {}", cached.display_name.as_deref().unwrap_or("(none)"));
            println!("     Model: {}", cached.model.as_deref().unwrap_or("(none)"));

            if let Some(create_time) = &cached.create_time {
                println!("     Created: {}", create_time);
            }

            if let Some(expire_time) = &cached.expire_time {
                println!("     Expires: {}", expire_time);
            }

            if let Some(usage) = &cached.usage_metadata {
                println!("     Total tokens: {}", usage.total_token_count);
            }
        }
    } else {
        println!("   No cached content found");
    }

    if let Some(next_page_token) = &response.next_page_token {
        println!("\n   More cached content available");
    }
    */

    println!("\n   Note: Lists all cached content with pagination support.");

    Ok(())
}

/// Example: Use cached content in generation.
async fn example_use_cached_content(client: &GeminiClientImpl, cached_name: &str) -> Result<(), Box<dyn std::error::Error>> {
    println!("2. Using cached content in generation...");
    println!("   Cached content: {}", cached_name);

    // Create a request that uses the cached content
    let request = GenerateContentRequest {
        contents: vec![
            Content {
                role: Some(Role::User),
                parts: vec![
                    Part::Text {
                        text: "Explain how neural networks work.".to_string(),
                    }
                ],
            }
        ],
        generation_config: None,
        safety_settings: None,
        tools: None,
        tool_config: None,
        system_instruction: None,
        cached_content: Some(cached_name.to_string()),
    };

    println!("   ✓ Request prepared with cached content reference");

    // Note: ContentService not yet implemented
    // Uncomment when available:
    /*
    let response = client.content()
        .generate("gemini-1.5-pro", request)
        .await?;

    println!("\n   ✓ Content generated successfully");

    if let Some(candidates) = &response.candidates {
        if let Some(candidate) = candidates.first() {
            if let Some(content) = &candidate.content {
                println!("\n   Response:");
                for part in &content.parts {
                    if let Part::Text { text } = part {
                        println!("   {}", text);
                    }
                }
            }
        }
    }

    // Display token savings
    if let Some(usage) = &response.usage_metadata {
        println!("\n   Token usage:");
        println!("     Prompt tokens: {}", usage.prompt_token_count);
        println!("     Completion tokens: {}", usage.candidates_token_count.unwrap_or(0));
        println!("     Total tokens: {}", usage.total_token_count);

        if let Some(cached_tokens) = usage.cached_content_token_count {
            println!("     Cached tokens: {} (not charged!)", cached_tokens);

            let savings_pct = if usage.total_token_count > 0 {
                (cached_tokens as f64 / usage.total_token_count as f64) * 100.0
            } else {
                0.0
            };
            println!("     Savings: {:.1}%", savings_pct);
        }
    }
    */

    println!("\n   Benefits of cached content:");
    println!("     • Reduced latency - cached content loads faster");
    println!("     • Lower costs - cached tokens are not charged");
    println!("     • Consistent context - same system instruction across requests");

    Ok(())
}

/// Example: Update cached content expiration.
async fn example_update_cached_content(client: &GeminiClientImpl, cached_name: &str) -> Result<(), Box<dyn std::error::Error>> {
    println!("2. Updating cached content expiration...");
    println!("   Cached content: {}", cached_name);

    let update_request = UpdateCachedContentRequest {
        ttl: Some("7200s".to_string()), // Extend to 2 hours
        expire_time: None,
    };

    println!("   New TTL: 7200s (2 hours)");

    // Note: CachedContentService not yet implemented
    // Uncomment when available:
    /*
    let updated_cached = client.cached_content()
        .update(cached_name, update_request)
        .await?;

    println!("\n   ✓ Cached content updated successfully");
    println!("   Name: {}", updated_cached.name);
    println!("   New expire time: {}", updated_cached.expire_time.unwrap_or_default());
    */

    println!("\n   Note: You can extend the TTL to keep content cached longer.");
    println!("   Maximum TTL varies by model and configuration.");

    Ok(())
}

/// Example: Delete cached content.
async fn example_delete_cached_content(client: &GeminiClientImpl, cached_name: &str) -> Result<(), Box<dyn std::error::Error>> {
    println!("2. Deleting cached content...");
    println!("   Cached content: {}", cached_name);

    // Note: CachedContentService not yet implemented
    // Uncomment when available:
    /*
    client.cached_content()
        .delete(cached_name)
        .await?;

    println!("\n   ✓ Cached content deleted successfully");

    // Verify deletion
    match client.cached_content().get(cached_name).await {
        Ok(_) => {
            println!("   ⚠ Warning: Cached content still exists after deletion");
        }
        Err(_) => {
            println!("   ✓ Confirmed: Cached content no longer exists");
        }
    }
    */

    println!("\n   Note: Deleted cached content cannot be recovered.");
    println!("   Cached content is automatically deleted after expiration.");

    println!("\n=== Best Practices for Cached Content ===");
    println!("  • Cache large, reusable system instructions or context");
    println!("  • Set appropriate TTL based on your use case");
    println!("  • Monitor cached token usage to understand savings");
    println!("  • Delete cached content when no longer needed");
    println!("  • Use descriptive display names for easier management");

    Ok(())
}
